parquet/src/main/java/org/apache/iceberg/parquet/ParquetAvro.java

```diff
@@ -252,7 +253,7 @@ public Schema record(Schema record, List<String> names, List<Schema> types) {
 
         newFields.add(copyField(field, type));
 
-        if (field.schema() != type) {
+        if (!Objects.equal(field.schema(), type)) {
```

Comment:
This was intentionally checking object identity because we want to rebuild if anything about the object changed, not just the fields that are used for equality. Part of the rationale is that we don't necessarily want to assume that `equals` is strict for types. But since we do have strict equality that checks all fields (including doc) I think this is fine.